jax/util.py

```diff
@@ -242,3 +244,69 @@ def wrap_name(name, transform_name):
 
 def extend_name_stack(stack, name=''):
   return stack + name + '/'
+
+
+class ContextVar:
+  """Like contextvars.ContextVar, but implemented with threading.local()."""
+  # TODO(shoyer): remove in favor of contextvars when JAX requires Python 3.7+
+
+  def __init__(self, name, *, default):
+    self.name = name
+    self.default = default
+    self._state = threading.local()
+    self._state.value = default
+    self._lock = threading.Lock()
+
+  def get(self):
+    return getattr(self._state, 'value', self.default)
+
+  def set(self, value):
+    with self._lock:
+      old_value = self.get()
+      self._state.value = value
+    return old_value
+
+  def reset(self, token):
+    self._state.value = token
+
+
+_OVERRIDES = {}
+
+
+def overrideable(name: str):
+  """Make an internal JAX function overrideable."""
+  def decorator(fun: Callable):
+    _OVERRIDES[name] = ContextVar(name, default=fun)
+    @functools.wraps(fun)
+    def wrapper(*args, **kwargs):
+      impl = _OVERRIDES[name].get()
+      with override_context({name: fun}):
```

Comment:
Is this nested `override_context` need for the case when the override wants to call the original function?